src/snapshot-editor/src/main.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use clap::{Parser, Subcommand};
+
+#[cfg(target_arch = "aarch64")]
+mod edit_vmstate;
+#[cfg(target_arch = "aarch64")]
+mod info;
```

Comment:
You can use [std::compile_error](https://doc.rust-lang.org/std/macro.compile_error.html) to avoid doing the `#[cfg(target_arch = "aarch64)]` thing multiple times. It is quite awkward otherwise that we produce a binary that does nothing on x86.

Reply:
yup, exactly. And then you can get rid of all the other directives (target_arch and unused)

Reply:
wait, this will not work, because when we build with `./tool/devtool build` it build everything, including `snapshot-editor`. And on x86 this build will fail. And CI will fail.

Reply:
I don't think this is worth it. There will be a follow up PR with additional functionality available on both x86 ant aarch64, so x86 version will be not useful only for short period of time.

Reply:
It doesn't look too complicated to me:

```diff
diff --git a/tools/release.sh b/tools/release.sh
index 1c84e510..2fb55039 100755
--- a/tools/release.sh
+++ b/tools/release.sh
@@ -60,6 +60,7 @@ function get-firecracker-version {
 # defaults
 LIBC=musl
 PROFILE=dev
+ARCH=$(uname -m)
 MAKE_RELEASE=

 #### Option parsing
@@ -130,7 +131,12 @@ fi

 say "Building version=$VERSION, profile=$PROFILE, target=$CARGO_TARGET..."
 # shellcheck disable=SC2086
-cargo build --target "$CARGO_TARGET" $CARGO_OPTS --workspace
+if [[ $ARCH != "aarch64" ]]; then
+    cargo build --target "$CARGO_TARGET" $CARGO_OPTS --workspace --exclude snapshot-editor
+else
+    cargo build --target "$CARGO_TARGET" $CARGO_OPTS --workspace
+fi
+

 say "Binaries placed under $CARGO_TARGET_DIR"
```

Reply:
The idea here is that this change will be reverted when x86 functionality will be added. Most `#[cfg(target_arch = "aarch64")]` will be removed anyway.

Reply:
I think the change in devtool makes sense, because why would we ship a binary to our customers that does nothing. It would just be confusing. If we do the devtool change, I dont think we need to change anything about the conditional compilation: We dont ship it, so it doesnt matter if it builds and produces a binary that does nothing, or if it doesnt even build.

Reply:
I know. I 'd just like to have a way to explicitly not build the tool for x86 until it actually does something on x86.

Reply:
Ok I added changes to the `release.sh`

Reply:
Well.. CI still fails because it still tries to build x86 version of `snapshot-editor`. I would prefer to have
 `#[cfg(target_arch = "aarch64")]` rather then `std::compile_error!()` in this case.

Reply:
Reverted some changes. Left `#[cfg(target_arch = "aarch64")]` only in `main.rs`

Reply:
I'd still like to not ship a binary that does something in the release for x86 (this implies fixing the tests), but I'd like to hear as well from others, @pb8o, @roypat

Reply:
This patch should fix the CI, btw: 

```diff
diff --git a/tests/host_tools/cargo_build.py b/tests/host_tools/cargo_build.py
index dffe0479..67af0e00 100644
--- a/tests/host_tools/cargo_build.py
+++ b/tests/host_tools/cargo_build.py
@@ -63,6 +63,8 @@ def cargo_test(path, extra_args=""):
         "RUST_BACKTRACE": 1,
         "RUSTFLAGS": get_rustflags(),
     }
+    if platform.machine() != "aarch64":
+        extra_args += "--exclude snapshot-editor"
     cargo("test", extra_args + " --all --no-fail-fast", env=env)


diff --git a/tests/integration_tests/build/test_clippy.py b/tests/integration_tests/build/test_clippy.py
index dc4ae415..7dafd797 100644
--- a/tests/integration_tests/build/test_clippy.py
+++ b/tests/integration_tests/build/test_clippy.py
@@ -22,4 +22,5 @@ def test_rust_clippy(target):
     """
     Test that clippy does not generate any errors/warnings.
     """
-    cargo("clippy", f"--target {target} --all --profile test", "-D warnings")
+    extra_args = "--exclude snapshot-editor" if platform.machine() != "aarch64" else ""
+    cargo("clippy", f"--target {target} --all --profile test {extra_args}", "-D warnings")
diff --git a/tests/integration_tests/build/test_coverage.py b/tests/integration_tests/build/test_coverage.py
index ac3af0b2..96714159 100644
--- a/tests/integration_tests/build/test_coverage.py
+++ b/tests/integration_tests/build/test_coverage.py
@@ -42,7 +42,7 @@ def test_coverage(monkeypatch):
     # Generate test profiles.
     cargo(
         "test",
-        f"--all --target {TARGET}",
+        f"--all --target {TARGET} --exclude snapshot-editor",
         "--test-threads=1",
         env={
             "RUSTFLAGS": "-Cinstrument-coverage",
diff --git a/tests/integration_tests/build/test_unittests.py b/tests/integration_tests/build/test_unittests.py
index 8bcb80cb..79f42bcd 100644
--- a/tests/integration_tests/build/test_unittests.py
+++ b/tests/integration_tests/build/test_unittests.py
@@ -21,10 +21,10 @@ def test_unittests(test_fc_session_root_path):
     Run unit and doc tests for all supported targets.
     """
     extra_args = "--target {} ".format(TARGET)
-
     host.cargo_test(test_fc_session_root_path, extra_args=extra_args)


 def test_benchmarks_compile():
     """Checks that all benchmarks compile"""
-    host.cargo("bench", f"--all --no-run --target {TARGET}")
+    extra_args = "--exclude snapshot-editor" if platform.machine() != "aarch64" else ""
+    host.cargo("bench", f"--all --no-run --target {TARGET} {extra_args}")
```

Reply:
But I thought with the patch to release.sh, we are no longer shipping a binary for x86_64 as part of the release?

as an aside, i think the patch for the CI can do directly into `def cargo(...)` in host_tools.py, that'd make for a smaller change to later revert.

Reply:
true, we can keep the `release.sh` changes and we're fine with the release binary. But we ship source code as well, so someone could get the code, build the tool and wait for it to do something on x86. It just sounds weird (apart from it being ugly). The CI changes are small enough (good catch Patrick about the `host.cargo` command) to just do them and then revert a single line once we add x86 functionality.

Reply:
If everybody else is fine with it, I won't block the PR for this.

Reply:
@roypat @pb8o could you chime in? this is the last open blocker